1.8/tutorial/resources.md

```diff
@@ -0,0 +1,83 @@
+      * Too few resources in cluster
+          * `dcos marathon app update /dcos-101/app2 instances=100` (potentially more if you have a large cluster)
+          * Use `dcos marathon app list` to check that the `scale` deployment is stuck
+          * `dcos marathon deployment list`
+          * Check marathon logs: TODO Discuss the easiest way to get marathon logs. `dcos service log marathon` does not work...
+          * The problem here is that there no matching resources (there might be for example resources left for the public-slace role, but not for *.
+          * Solution: add nodes are scale app back to level at which resources are available `dcos marathon app update /dcos-101/app2 --force instances=1`. Note you have to use the `--force` flag here as the previous deployment is ongoing.
+      * Too few resources on single node. Each app is started on a single node, hence resources must fit onto a single node.
+          * `dcos marathon app update /dcos-101/app2 cpus=100`
+          * Use `dcos marathon app list` to check that the `restart` deployment is stuck
+          * `dcos marathon deployment list`
+          * Check marathon logs: TODO Discuss the easiest way to get marathon logs
+          * The problem here is that there no resource offer being large enough to match the offer.
+          * Solution: Privision larger nodes or scale app down to level where it fits onto the free resources on a single node `dcos marathon app update /dcos-101/app2 --force cpus=1`. Note that we have to use the force flag again.
+   * Debugging resource isolation
+      * What happens if an app tries to use more resources then it is allocated. Most common problem is memory consumption in conjunction with JVM based applications.
+      * Deploy [memory eater app](https://github.com/joerg84/dcos-101/blob/master/oomApp/oomApp.go)
+          * `wget https://raw.githubusercontent.com/joerg84/dcos-101/master/oomApp/oomApp.json`
+          * `dcos marathon app add oomApp.json`
+          * You will see it restarting over and over again
+          * Check marathon Log. Potentially we can see the oom already here.
+          * Problem here: Kernel is killing the app which not necessarly visible to DC/OS.
+          * SSH to an agent where the app has been running `dcos node ssh --master-proxy --mesos-id=$(dcos task oom-app --json | jq -r '.[] | .slave_id')'
+          * Check Kernel log `journalctl -f _TRANSPORT=kernel`
+          * Here we see something like ` Memory cgroup out of memory: Kill process 10106 (oomApp) score 925 or sacrifice child; Killed process 10390 (oomApp) total-vm:3744760kB, anon-rss:60816kB, file-rss:1240kB, shmem-rss:0kB`
+          * Solution: Check app for correct behavior and/or increase the allocated memory
+          * Remove app `dcos marathon app remove /dcoc-101/oom-app`
+
+
+# Outcome
+We explored the difference between allocated and used resources. Furthermore, we learned how to debug issues related to resources.
```

Comment:
Congratulations! You've now learned how to deploy apps to DC/OS, network those apps, expose them to the outside of the cluster with a load-balancer, scale them, and debug potential resource issues! You're practically a pro!